### PR TITLE
Support NullMarked and NullUnmarked Annotations

### PR DESCRIPTION
**Pull request by HarrisL2**
Based on #26456 

Adds support for NullMarked and NullUnmarked. Inside Java source, symbols with NullMarked in scope will be non-null, and NullUnmarked scopes remain implicit.

## Have you relied on LLM-based tools in this contribution?

Yes, and I checked the output by hand and writing the corresponding tests.

## How was the solution tested?

New automated tests

### CODE REVIEW

File: tests/explicit-nulls/neg/nullmarked-packageinfo-separate/S_1.scala
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked-packageinfo-separate/S_1.scala:1]: Any scala tests need to be tested with `-Yno-flexible-types` so that it errors on the default unannotated case

File: tests/explicit-nulls/neg/nullmarked-packageinfo/S.scala
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked-packageinfo/S.scala:1]: Any tests must be compiled with -Yno-flexible-types

File: tests/explicit-nulls/neg/nullmarked-separate/S_3.scala
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked-separate/S_3.scala:1]: Any tests must be compiled with -Yno-flexible-types

File: tests/explicit-nulls/neg/nullmarked/S.scala
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked/S.scala:1]: Any tests must be compiled with -Yno-flexible-types

**Review by TheDrawingCoder-Gamer — CHANGES_REQUESTED** [review]

File: tests/explicit-nulls/neg/nullmarked-packageinfo-separate/J.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    public @Nullable String getNullable() {
+        return null;
+    }
+
+    // Unannotated parameter -> non-null String.
+    public void set(String s) {
+    }
+
+    // `@NullUnmarked` overrides the package marking for this method.
+    @NullUnmarked
```
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked-packageinfo-separate/J.java:23]: We should probably test all forms of null unmarking (methods, _named_ classes (including static and non static), and packages)

static classes tripped my solution up due to them getting skipped on certain symbol extension methods

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala [hunk 2]
```diff
@@ -94,9 +100,37 @@ object ImplicitNullInterop:
...
+  /** The scope marking declared directly on `owner`, if any: `Some(true)` for `@NullMarked`,
+   *  `Some(false)` for `@NullUnmarked`, `None` if `owner` declares neither.
+   *
+   *  For packages the marker is placed on `package-info` (from `package-info.java` /
+   *  `package-info.class`), which is loaded as a synthetic `package-info` member of the package.
+   *  We force that member to read its annotations (safe: it only depends on the annotation
+   *  classes). For all other owners we use `unforcedAnnotation` to avoid forcing symbols that may
+   *  still be under construction during classfile loading / unpickling.
+   */
+  private def ownerNullMarking(owner: Symbol)(using Context): Option[Boolean] =
```
Comment (TheDrawingCoder-Gamer) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:124]: jSpecify has a weird edge case where if someone defines something with both nullmarked _and_ null unmarked, it's interpreted like it has neither. This code doesn't seem to implement that

File: tests/explicit-nulls/neg/nullmarked-packageinfo/package-info.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// The whole package is `@NullMarked` via `package-info.java`, so unannotated reference types
+// in this package are non-null. JavaParsers must read this package-level annotation.
+@org.jspecify.annotations.NullMarked
```
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked-packageinfo/package-info.java]: We should also test imported annotations as many java users write something like this for their package info:

```java
@NullMarked
package nmpkg;

import org.jspecify.annotations.*;
```

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala (hunk 2 above)
Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:124]: Wow, that is indeed a weird edge case, where is this specified?

Comment (TheDrawingCoder-Gamer) [reply compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:124]: Here:
<https://jspecify.dev/docs/spec/#null-marked-scope>
> Two of the rules here are worth further discussion.
>
> First: If a given declaration is annotated with both @NullMarked and @NullUnmarked, these rules behave as if neither annotation is present.

File: tests/explicit-nulls/neg/nullmarked-separate/J_2.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    // `@NullUnmarked` on a nested class restores the implicit-nulls default for its members.
+    @NullUnmarked
+    public static class Inner {
+
+        // Unmarked -> String | Null.
+        public String innerGet() {
+            return "";
+        }
+
+        // Unmarked parameter -> nullable.
```
Comment (TheDrawingCoder-Gamer) [inline tests/explicit-nulls/neg/nullmarked-separate/J_2.java:50]: 🤓 nullable isn't technically correct, the correct typing is flexible type (denoted internally as `String?`)

Of course without flexible types yes, it is indeed nullable

this is nitpicking but maybe "unspecified nullability" would be more accurate to both scala & jSpecify?

Comment (HarrisL2) [reply tests/explicit-nulls/neg/nullmarked-separate/J_2.java:50]: fair, within this test it is nullable though, since we're in explicit nulls and without flexible types

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 5]
```diff
@@ -1225,6 +1225,24 @@ class Definitions {
...
+
+  @tu lazy val NullUnmarkedAnnots: List[ClassSymbol] = getClassesIfDefined(
+    "org.jspecify.annotations.NullUnmarked" :: Nil)
+
+  /** The name of the synthetic member that carries a package's annotations, loaded from
+   *  `package-info.java` / `package-info.class` (`package-info` is not a legal Java identifier,
+   *  so it never collides with a real member). Used by the explicit-nulls interop to read a
+   *  package-level `@NullMarked` / `@NullUnmarked`.
+   */
+  val PackageInfoName: TypeName = typeName("package-info")
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1244]: Move to `JavaNames` in `StdNames.scala`

**Review by olhotak — APPROVED** [review]
Otherwise LGTM, but @noti0na1 should also have a look.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 5 above)
Comment (olhotak) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:1244]: I withdraw this suggestion. It looks like those `JavaNames` are not used much anymore elsewhere in the compiler.

**Review by TheDrawingCoder-Gamer — APPROVED** [review]
also LGTM

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala [hunk 6]
```diff
@@ -94,9 +99,43 @@ object ImplicitNullInterop:
...
+      ambient = ambient,
       state = NullMapState(resultTypeMode, currentTypeMode)
     )(tp)
 
+  /** Is `sym` in a JSpecify `@NullMarked` scope? We walk the owner chain (starting at `sym`
+   *  itself) and let the nearest scope marking win: `@NullMarked` enables the non-null default,
+   *  `@NullUnmarked` re-enables the implicit-nulls default.
+   */
+  private def isNullMarked(sym: Symbol)(using Context): Boolean =
+    sym.ownersIterator.map(ownerNullMarking).collectFirst { case Some(marked) => marked }.getOrElse(false)
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:111]: According to the document (https://jspecify.dev/docs/api/org/jspecify/annotations/NullMarked.html):

> `@NullMarked` and `@NullUnmarked` ... To apply these annotations to an entire (single) package, create a [package-info.java](https://docs.oracle.com/javase/specs/jls/se26/html/jls-7.html#jls-7.4.1) file and annotate the package declaration there. This annotation has no effect on "subpackages".

We may want to stop the iteration at package level.

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala [hunk 7]
```diff
@@ -94,9 +99,43 @@ object ImplicitNullInterop:
...
+   *  declaration carrying *both* markers behaves as if it carried neither, so we return `None` and
+   *  let an enclosing scope decide.
+   *
+   *  For packages the marker is placed on `package-info` (from `package-info.java` /
+   *  `package-info.class`), which is loaded as a synthetic `package-info` member of the package.
+   *  We force that member to read its annotations (safe: it only depends on the annotation
+   *  classes). For all other owners we use `unforcedAnnotation` to avoid forcing symbols that may
+   *  still be under construction during classfile loading / unpickling.
+   */
+  private def ownerNullMarking(owner: Symbol)(using Context): Option[Boolean] =
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:124]: Do we want to also handle `module-info`?

File: compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala [hunk 8]
```diff
@@ -115,22 +154,29 @@ object ImplicitNullInterop:
...
    *  coming from Scala code compiled without explicit nulls, this adds `| Null` or `FlexibleType` in the
    *  right places to make nullability explicit in a conservative way (without forcing incomplete symbols).
    *
    *  @param javaDefined  whether the type is from Java source; we always nullify type param refs from Java
+   *  @param ambient      the default mode for unannotated positions: `Default` normally, or `Skip` inside
+   *                      a JSpecify `@NullMarked` scope (where unannotated reference types are non-null).
    *  @param state        mutable nullification state tracking the current mode for the result type
    *                      (`resultTypeMode`) and the current nesting level (`currentTypeMode`).
    */
   private class ImplicitNullMap(
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/ImplicitNullInterop.scala:177]: I saw the spec has some special rules for wildcards and type parameters. Are they supported by this PR? If not, I'm not sure if we want to support them?

**Review by noti0na1 — CHANGES_REQUESTED** [review]
The core change looks good to me. The test suit need to be extended for:
* sub packages;
* a jar, if possible;
* generics;
* a package marked by both annotations.
